test/performance/odp_ipsecfwd.c

```diff
@@ -986,110 +989,157 @@ static inline const pktio_t *lookup_and_apply(odp_packet_t pkt, odph_table_t fwd
+			continue;
+		}
+
+		if (odp_unlikely(!pktio->send_fn(pktio, hash % pktio->num_tx_qs, pkt))) {
+			odp_packet_free(pkt);
+			continue;
+		}
+
+		++num_procd;
 	}
 
-	return true;
+	return num_procd;
 }
 
-static inline void process_packet_out(odp_packet_t pkt, odph_table_t fwd_tbl, stats_t *stats)
+static inline void process_packets_out(odp_packet_t pkts[], int num, odph_table_t fwd_tbl,
+				       stats_t *stats)
 {
-	odp_ipsec_sa_t *sa;
+	odp_packet_t pkt, pkts_ips[MAX_BURST], pkts_fwd[MAX_BURST];
+	odp_ipsec_sa_t *sa, sas[MAX_BURST];
+	int num_pkts_ips = 0, num_pkts_fwd = 0, num_procd;
+
+	for (int i = 0; i < num; ++i) {
+		pkt = pkts[i];
+		sa = get_out_sa(pkt);
+
+		if (sa != NULL) {
+			sas[num_pkts_ips] = *sa;
+			pkts_ips[num_pkts_ips] = pkt;
+			++num_pkts_ips;
+		} else {
+			pkts_fwd[num_pkts_fwd++] = pkt;
+		}
+	}
 
-	sa = get_out_sa(pkt);
+	if (num_pkts_ips > 0) {
+		num_procd = process_ipsec_out(pkts, sas, num_pkts_ips);
 
-	if (sa != NULL) {
-		if (odp_unlikely(!process_ipsec_out(pkt, sa))) {
-			++stats->ipsec_out_errs;
-			goto err;
+		if (odp_unlikely(num_procd < num_pkts_ips)) {
+			num_procd = num_procd < 0 ? 0 : num_procd;
+			stats->ipsec_out_errs += num_pkts_ips - num_procd;
+			odp_packet_free_multi(&pkts_ips[num_procd], num_pkts_ips - num_procd);
 		}
-	} else {
-		if (!forward_packet(pkt, fwd_tbl, stats))
-			goto err;
-
-		++stats->fwd_pkts;
 	}
 
-	return;
-
-err:
-	odp_packet_free(pkt);
+	if (num_pkts_fwd > 0) {
+		num_procd = forward_packets(pkts_fwd, num_pkts_fwd, fwd_tbl);
+		stats->discards += num_pkts_fwd - num_procd;
+		stats->fwd_pkts += num_procd;
+	}
 }
 
-static inline void process_packet_in(odp_packet_t pkt, odph_table_t fwd_tbl, stats_t *stats)
+static inline void process_packets_in(odp_packet_t pkts[], int num, odph_table_t fwd_tbl,
+				      stats_t *stats)
 {
-	odp_ipsec_sa_t *sa;
+	odp_packet_t pkt, pkts_ips[MAX_BURST], pkts_out[MAX_BURST];
+	odp_ipsec_sa_t *sa, sas[MAX_BURST];
+	int num_pkts_ips = 0, num_pkts_out = 0, num_procd;
 
-	if (odp_unlikely(odp_packet_has_error(pkt))) {
-		++stats->discards;
-		goto err;
-	}
+	for (int i = 0; i < num; ++i) {
+		pkt = pkts[i];
+
+		if (odp_unlikely(odp_packet_has_error(pkt))) {
+			++stats->discards;
+			odp_packet_free(pkt);
+			continue;
+		}
 
-	sa = get_in_sa(pkt);
+		sa = get_in_sa(pkt);
 
-	if (sa != NULL) {
-		if (odp_unlikely(!process_ipsec_in(pkt, sa))) {
-			++stats->ipsec_in_errs;
-			goto err;
+		if (sa != NULL) {
+			sas[num_pkts_ips] = *sa;
+			pkts_ips[num_pkts_ips] = pkt;
+			++num_pkts_ips;
+		} else {
+			pkts_out[num_pkts_out++] = pkt;
 		}
-	} else {
-		process_packet_out(pkt, fwd_tbl, stats);
 	}
 
-	return;
+	if (num_pkts_ips > 0) {
+		num_procd = process_ipsec_in(pkts_ips, sas, num_pkts_ips);
+
+		if (odp_unlikely(num_procd < num_pkts_ips)) {
+			num_procd = num_procd < 0 ? 0 : num_procd;
+			stats->ipsec_in_errs += num_pkts_ips - num_procd;
+			odp_packet_free_multi(&pkts_ips[num_procd], num_pkts_ips - num_procd);
```

Comment:
Maybe the enqueuing should not be given up so easily. An ODP implementation may have a limited maximum burst size that it can handle so that it only consumes at most that many packets even if the offered burst size is longer. This is problematic especially with the fixed burst size in this test app.

Enqueuing of the remaining packets could be retried as long as odp_ipsec_XXX_enq() consumed some packets.

Reply:
Yes, good point. Added a retry logic in v2.

Reply:
The logic is good but the macro is pretty ugly and has the usual macro problem with arguments getting evaluated multiple times. An inline function would be much better, but you could also make single inlined process_ipsec funtion that would handle both the inbound and outbound case based on a parameter passed and the compiler would just optimize the constant parameter and the conditionals away.

e.g. process_ipsec_in() could then be like this:
static inline int process_ipsec_in(odp_packet_t pkts[], const odp_ipsec_sa_t sas[], int num)
{
	return process_ipsec(DIR_IN, pkts, sas, num);
}

Reply:
Sure, macro is a bit ugly though a single function here doesn't work without type erasure (which is also ugly) as operation parameters are of different type. "Umbrella" functions that do different things based on input boolean flags are not that nice either.

Maybe I'll just inline the `while()` logic to both in and out processing functions.

Also just noticed a bug: need to also adjust the `num_sa` and `sa` fields of operation params according to processed packets.

Reply:
> Maybe I'll just inline the while() logic to both in and out processing functions.

I think that would be fine too.